pkg/middleware/auth.go

```diff
@@ -90,6 +101,95 @@ func (c JWTVerifier) CheckToken(token string) (*jwt.Token, error) {
 	return tk, nil
 }
 
+// JWTVerifierURL checks token with JWT against JWKS, implements Verifier
+type JWTVerifierURL struct {
+	jwksURL string
+
+	parser   *jwt.Parser
+	jwksKeys *jwk.Set
+	jwksLock sync.RWMutex
+}
+
+func (c *JWTVerifierURL) startTicker(ticker time.Duration) {
+
+	for range time.Tick(ticker) {
+		err := c.updateKeySet()
+		if err != nil {
+			log.Warn().Msg("unable to update JWKS key set")
+		}
+	}
+}
+
+func (c *JWTVerifierURL) updateKeySet() error {
+
+	c.jwksLock.Lock()
+	defer c.jwksLock.Unlock()
+
+	set, err := jwk.Fetch(c.jwksURL)
+	if err != nil {
+		return err
+	}
+	c.jwksKeys = set
+	return nil
+}
+
+func (c *JWTVerifierURL) getKeySet() *jwk.Set {
+	c.jwksLock.RLock()
+	defer c.jwksLock.RUnlock()
+	return c.jwksKeys
+
+}
+
+// NewJWTVerifierURL creates JWTVerifierURL with JWKS URL
+func NewJWTVerifierURL(jwksURL string, updateInterval time.Duration) *JWTVerifierURL {
+
+	http.DefaultClient = &http.Client{Timeout: 10 * time.Second}
+	jwtVerifierURL := JWTVerifierURL{jwksURL: jwksURL, parser: new(jwt.Parser)}
+	err := jwtVerifierURL.updateKeySet()
+
+	if err != nil {
+		return &JWTVerifierURL{}
+	}
+
+	go jwtVerifierURL.startTicker(updateInterval)
```

Comment:
Should `NewJWTVerifierURL` accept a context, and cancel its ticker/ in-flight requests when the context is cancelled?

Reply:
I don't think it's needed. it's just the updater, and it will exit when the main thread exits. We don't have to wait for the complete update.

Reply:
From [the docs](https://golang.org/pkg/time/#Tick)

> Tick is a convenience wrapper for NewTicker providing access to the ticking channel only. While Tick is useful for clients that have no need to shut down the Ticker, be aware that without a way to shut it down the underlying Ticker cannot be recovered by the garbage collector; it "leaks". Unlike NewTicker, Tick will return nil if d <= 0.